src/mono/wasm/runtime/binding_support.js

```diff
@@ -938,7 +1054,339 @@ var BindingSupportLib = {
+				}
+			}
+
+			return this._struct_unboxer_cache.get (typePtr);
+		},
+
+		_unbox_struct_rooted: function (unbox_buffer, mono_obj) {
+			var objSize = Module.HEAP32[(unbox_buffer / 4) | 0];
+			var classPtr = Module.HEAP32[((unbox_buffer / 4) | 0) + 1];
+			var dataOffset = unbox_buffer + 8;
+			if (!classPtr)
+				throw new Error("classPtr is null or undefined");
+
+			// console.log (`objSize ${objSize} classPtr ${classPtr} dataOffset ${dataOffset}`);
+
+			var typePtr = this.mono_wasm_class_get_type(classPtr);
+			var unboxer = this._get_struct_unboxer_for_type(typePtr);
+			if (!unboxer)
+				throw new Error (`No CustomJavaScriptMarshaler found for struct type ${this._get_type_name(typePtr)}`);
+
+			// FIXME: Pass a ReadOnlySpan or ReadOnlyMemory instead of a bare pointer
+			return unboxer (dataOffset);
+		},
+
+		_compile_pre_filter: function (typePtr, boundConverter, js) {
+			if (!js)
+				return boundConverter;
+			
+			var closure = { 
+				MONO: MONO,
+				BINDING: this,
+				typePtr: typePtr, 
+				// (js_obj, method, parmIdx) => value
+				boundConverter: boundConverter 
+			};
+			var body = [
+				"var filteredValue = null;",
+				// "console.log(`preFilter(${value})`);",
+				`{ filteredValue = ${js}; }`,
+				// "console.log(`preFilter === ${filteredValue}`);",
+				"var convertedResult = boundConverter (filteredValue, method, parmIdx);",
+				// "console.log(`convertedResult === ${convertedResult}`);",
+				"return convertedResult;"
+			];
+			
+			var bodyJs = body.join ("\r\n");
+			var result = this._create_named_function(
+				"pre_filtered_converter_for_type" + typePtr, 
+				["value", "method", "parmIdx"], bodyJs, closure
+			);
+
+			// console.log("compile result", result);
+			return result;
+		},
+
+		_pick_automatic_converter_for_user_type: function (methodPtr, args_marshal, typePtr) {
+			if (!typePtr)
+				throw new Error("typePtr is null or undefined");
+
+			if (!this._automatic_converter_table)
+				this._automatic_converter_table = new Map ();
+			if (!this._automatic_converter_table.has (typePtr)) {
+				var info = this._get_custom_marshaler_info_for_type (typePtr);
+				// HACK
+				if (!info)
+					info = {};
+				if (info.error)
+					console.error(`Error while configuring automatic converter for type ${this._get_type_name(typePtr)}: ${info.error}`);
+
+				var preFilter = info.preFilter;
+
+				// console.log ("preFilter", preFilter);
+
+				var convMethod = info.inputPtr;
+				if (!convMethod) {
+					if (info.typePtr)
+						console.error(`Automatic converter for type ${this._get_type_name(typePtr)} has no suitable FromJavaScript method`);
+					this._automatic_converter_table.set (typePtr, null);
+					return null;
+				}
+
+				// FIXME
+				var sigInfo = this.get_method_signature_info (0, convMethod);
+				// Return unboxed so it can go directly into the arguments list
+				var signature = this._pick_result_chara_for_marshal_type (sigInfo.parameters[0].marshalType) + "!";
```

Comment:
Why not do this in managed?

Reply:
This logic is all used in order to call managed methods, so I'm uncomfortable constantly calling managed methods in order to call managed methods. We can probably do it in this case though if you want - this bit of code predates moving some other stuff to managed (at the time it was done natively in driver.c, but I had to move it because our C APIs are busted).